auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java

```diff
@@ -20,6 +43,120 @@ public OIDCAuthCodeImpl(final AuthServiceConfig config) {
+    }
+
+    private String parseEmail() throws JSONException {
+        String emailAddress = "Unknown Email";
+        if (userIdentity != null) {
+            // get the users email
+            if (userIdentity.has(EMAIL) && userIdentity.getString(EMAIL).length() > 0) {
+                emailAddress = userIdentity.getString(EMAIL);
+            }
+        }
+        return emailAddress;
+    }
+
+    private Map<RoleKey, IRole> parseRoles() throws JSONException {
+        Map<RoleKey, IRole> roles = new HashMap<>();
+        if (userIdentity != null) {
+            Map<RoleKey, IRole> realmRoles = parseRealmRoles();
+            if (realmRoles != null) {
+                roles.putAll(realmRoles);
+            }
+            Map<RoleKey, IRole> clientRoles = parseClientRoles();
+            if (clientRoles != null) {
+                roles.putAll(clientRoles);
+            }
+        }
+        return roles;
+    }
+
+    private Map<RoleKey, IRole> parseRealmRoles() throws JSONException {
+        Map<RoleKey, IRole> realmRoles = new HashMap<>();
+        if (userIdentity.has(REALM) && userIdentity.getJSONObject(REALM).has(ROLES)) {
+            String tokenRealmRolesJSON = userIdentity.getJSONObject(REALM).getString(ROLES);
+
+            String realmRolesString = tokenRealmRolesJSON.substring(1, tokenRealmRolesJSON.length() - 1);
+            String roles[] = realmRolesString.split(COMMA);
+
+            for (String rolename : roles) {
+                RealmRole realmRole = new RealmRole(rolename);
+                realmRoles.put(new RoleKey(realmRole), realmRole);
+            }
+        }
+        return realmRoles;
+    }
+
+    private Map<RoleKey, IRole> parseClientRoles() throws JSONException {
+        Map<RoleKey, IRole> clientRoles = new HashMap<>();
+
+        AuthServiceConfig authConfig = this.getConfig();
+        JSONObject authJSON =  authConfig.toJSON();
+
+        if (authJSON.has(RESOURCE)) {
+            String initialClientID = authJSON.getJSONObject(RESOURCE).toString();
+
+            if (userIdentity.has(CLIENT) && userIdentity.getJSONObject(CLIENT).has(initialClientID)
+                    && userIdentity.getJSONObject(CLIENT).getJSONObject(initialClientID).has(ROLES)) {
+                String tokenClientRolesJSON = userIdentity.getJSONObject(CLIENT).getJSONObject(initialClientID).getString(ROLES);
+
+                String clientRolesString = tokenClientRolesJSON.substring(1, tokenClientRolesJSON.length() - 1);
+                String roles[] = clientRolesString.split(COMMA);
+
+                for (String rolename : roles) {
+                    ClientRole clientRole = new ClientRole(rolename);
+                    clientRoles.put(new RoleKey(clientRole), clientRole);
+                }
+            }
+        }
+        return clientRoles;
+    }
+
+    private JSONObject getIdentityInformation(final ICredential credential) throws JSONException {
+        String accessToken = credential.getAccessToken();
```

Comment:
A cast to `OIDCCredential` is needed here.
A check that `ICredential` is an instance of `OIDCCredential` would be a good idea. If it is not, an IllegalArgumentException should be raised